core/src/world/place/building/religious/shrine.rs

```diff
@@ -0,0 +1,238 @@
+        0..=2 => format!("The {} {}", descriptor(rng), place(rng)),
+        3..=5 => format!("{} of {}", place(rng), deity(rng)),
+        6 => format!("Place where the {} {}",pluralize(animal(rng)),action(rng)),
+        7 => format!("{} of the {} {}",place(rng),number(rng),pluralize(animal(rng))),
+        _ => unreachable!(),
+    }
+}
+
+//place of worship can be a building or a natural feature
+fn place(rng: &mut impl Rng) -> &'static str {
+    let choice = rng.gen_range(0..3);
+    if choice == 0 {
+        feature(rng)
+    } else {
+        building(rng)
+    }
+}
+
+//commonly worshipped places
+fn building(rng: &mut impl Rng) -> &'static str {
+    const BUILDINGS: &[&str] = &[
+        "Altar",
+        "Fane",
+        "Pagoda",
+        "Shrine",
+        "Cave",
+        "Tree",
+        "Figure",
+        "Gate",
+        "Monolith",
+        "Obelisk",
+        "Pagoda",
+        "Pillar",
+        "Pillars",
+        "Icon",
+    ];
+    BUILDINGS[rng.gen_range(0..BUILDINGS.len())]
+}
+
+//less common places of worship, typically natural formations
+fn feature(rng: &mut impl Rng) -> &'static str {
+    #[rustfmt::skip]
+    const FEATURES: &[&str] = 
+        &[
+        "Basin","Boulder","Cavern","Grove","Pond","Pool","Menhir",
+        "Grotto","Cenote", "Grove", "Tree", "Stones", "Cave"
+        ];
+    FEATURES[rng.gen_range(0..FEATURES.len())]
+}
+
+//DESCRIPTOR can be an ADJECTIVE, an ACTION, or another noun that fits well e.g. PHOENIX
+fn descriptor(rng: &mut impl Rng) -> String {
+    match rng.gen_range(0..3) {
+        0 => adjective(rng),
+        1 => gerund(action(rng)),
+        2 => noun(rng),
+        _ => unreachable!(),
+    }
+}
+
+//ADJECTIVE
+fn adjective(rng: &mut impl Rng) -> String {
+    #[rustfmt::skip]
+    const ADJECTIVES: &[&str] = &[
+        "Amaranthine","Ancestral","Ancient","Angelic","Argent","Astral","Azure",
+        "Blessed","Blue","Bright","Celestial","Corrupted","Dark","Devout",
+        "Divine","Elder","Eternal","Ethereal","Exalted","Foul","Golden","Guilty","Hallowed",
+        "Heavenly","Immortal","Impure","Ivory","Shining","Lucent","Pale","Primal","Putrid",
+        "Radiant","Red","Rusted","Sacred","Sanctified","Sanguine","Silver","Solemn","Tainted",
+        "Timeless","Tribal","True","Vile","White","Wicked",
+    ];
+    ADJECTIVES[rng.gen_range(0..ADJECTIVES.len())].to_string()
+}
+//ACTION
+fn action(rng: &mut impl Rng) -> String {
+    #[rustfmt::skip]
+    const ACTIONS: &[&str] = &[
+        "Dance","Whisper","Shiver","Rot","Rise","Fall","Laugh","Travel","Creep",
+        "Sing","Fade","Glow","Shine","Stand","Weep","Drown","Howl","Smile",
+    ];
+    ACTIONS[rng.gen_range(0..ACTIONS.len())].to_string()
+}
+
+fn gerund(verb: String) -> String {
+    let last_char = verb.chars().last().unwrap();
+    if last_char == 'e' {
+        format!("{}ing", &verb[..verb.len() - 1])
+    } else {
+        format!("{}ing", verb)
+    }
+}
+
+fn pluralize(noun: String) -> String {
```

Comment:
This and the `plural()` function in the `inn` mod can be merged into the `::utils` mod. It's going to get used in a lot of generators I think.

General Rust best practice is to use `&str` instead of `String` whenever possible in function signatures, although as this isn't a particularly hot code path, unnecessary copying isn't the end of the world. I may have gone overboard to avoid copying in `inn::plural()`.

Reply:
not sure I understand what's going on with a &str vs String in rust just yet :thinking: moved `pluralize()` over and added any missing cases from `plural()` and I'll combine then once I've got a handle on this bit

Reply:
Yeah, it's one of the harder Rust concepts to internalize in practice, particularly for those of us accustomed to garbage-collected languages. `&str` is just a pointer to a string stored elsewhere in memory, either statically in the application binary or in a `String` somewhere else. The [Rust Book](https://doc.rust-lang.org/book/ch04-03-slices.html) talks more about how and why they work.

But also: most higher-level languages just gratuitously copy strings around whenever they feel like it, so it's easy to fall down the rabbit hole of premature optimization here.